### PR TITLE
Changed the way loss curves and maps are computed and stored in event_based_risk

### PR DESCRIPTION
**Pull request by micheles**
Second part of https://github.com/gem/oq-engine/pull/2714. Loss curves are now computed on the fly for a given asset ID. Loss maps are stored as before, but the procedure to compute them is different, since the loss ratios are being stored differently, in the datastore and not in the .ext5 file. The procedure is still parallelized, but only in the controller node. This is fast enough for all known computations, and it is a step in the direction of removing the .ext5 file (https://github.com/gem/oq-engine/issues/2718).
Another big result is that now the datastore is pickleable, i.e. it can be sent to the workers.

Closes https://github.com/gem/oq-engine/issues/2707 and the first three points of https://github.com/gem/oq-engine/issues/2673.

The demos are green: https://ci.openquake.org/job/zdevel_oq-engine/2469

### CODE REVIEW

File: openquake/calculators/event_based_risk.py [hunk 1]
```diff
@@ -529,27 +553,34 @@ def save_results(self, allres, num_rlzs):
...
 
-    def save_losses(self, dic, offset=0):
+    def save_losses(self, dic, taskno, offset=0):
         """
         Save the event loss tables incrementally.
 
         :param dic:
-            dictionary with agglosses, asslosses, losses_by_taxon, avglosses
+            dictionary with agglosses, assratios, losses_by_taxon, avglosses
         :param offset:
```
Comment (ptormene) [inline openquake/calculators/event_based_risk.py]: Please add `taskno` to the docstring

File: openquake/calculators/event_based_risk.py [hunk 2]
```diff
@@ -529,27 +553,34 @@ def save_results(self, allres, num_rlzs):
...
             realization offset
         """
         aids = dic.pop('aids')
         agglosses = dic.pop('agglosses')
-        asslosses = dic.pop('asslosses')
+        assratios = dic.pop('assratios')
         losses_by_taxon = dic.pop('losses_by_taxon')
         avglosses = dic.pop('avglosses')
-        with self.monitor('saving event loss tables', autoflush=True):
+        lrs_idx = dic.pop('lrs_idx')
```
Comment (ptormene) [inline openquake/calculators/event_based_risk.py]: Please add `lrs_idx` to the description of the dictionary in the docstring

File: openquake/calculators/tests/event_based_risk_test.py [hunk 3]
```diff
@@ -74,21 +75,18 @@ def assert_stats_ok(self, pkg, job_ini, individual_curves='false'):
...
-            ('agg_curve-stats', 'xml'),
-        ]
+        ekeys = [('agg_curve-stats', 'xml')]
         for ekey in ekeys:
             for fname in export(ekey, self.calc.datastore):
                 self.assertEqualFiles(
                     'expected/%s' % strip_calc_id(fname), fname)
 
+        fnames = export(('loss_maps-stats', 'csv'), self.calc.datastore)
+        if REFERENCE_OS:
```
Comment (ptormene) [inline openquake/calculators/tests/event_based_risk_test.py]: Should we have the sanity check `assert fnames` also here and in all the other occurrences?

File: openquake/calculators/tests/event_based_risk_test.py [hunk 4]
```diff
@@ -173,7 +171,14 @@ def test_occupants(self):
                  out['agg_curve-rlzs', 'xml']
         self.assertEqual(len(fnames), 3)  # 2 loss_maps + 1 agg_curve
         for fname in fnames:
-            self.assertEqualFiles('expected/' + strip_calc_id(fname), fname)
+            self.assertEqualFiles('expected/' + strip_calc_id(fname),
+                                  fname, delta=1E-5)
+
+        fnames = export(('loss_maps-rlzs', 'csv'), self.calc.datastore)
+        if REFERENCE_OS:
```
Comment (ptormene) [inline openquake/calculators/tests/event_based_risk_test.py]: Same as above

File: openquake/risklib/scientific.py [hunk 5]
```diff
@@ -875,13 +876,13 @@ def survival(self, loss_ratios, probs):
...
-class CurveBuilder(object):
+class LossTypeCurveBuilder(object):
     """
     Build loss ratio curves. The loss ratios can be provided
     by the user or automatically generated (user_provided=False).
     The usage is something like this::
 
-      builder = CurveBuilder(loss_type, loss_ratios, ses_ratio,
+      builder = LossTypeCurveBuilder(loss_type, loss_ratios, ses_ratio,
                              user_provided=True)
```
Comment (ptormene) [inline openquake/risklib/scientific.py]: Please fix the indentation

File: openquake/risklib/scientific.py [hunk 6]
```diff
@@ -950,6 +951,144 @@ def __repr__(self):
...
+        """"
+        :param assets:
+            a list of assets
+        :param getter:
+            a :class:`openquake.risklib.riskinput.LossRatiosGetter` instance
+        :param rlzs:
+            a list of realizations
+        :param quantiles:
+            a list of quantiles
+        :returns:
```
Comment (ptormene) [inline openquake/risklib/scientific.py]: Please add the description of `mon`

File: openquake/calculators/tests/event_based_risk_test.py (hunk 3 above)
Comment (micheles) [inline openquake/calculators/tests/event_based_risk_test.py]: yes

File: openquake/calculators/event_based_risk.py (hunk 2 above)
Comment (micheles) [inline openquake/calculators/event_based_risk.py]: Done

File: openquake/calculators/tests/event_based_risk_test.py (hunk 4 above)
Comment (micheles) [inline openquake/calculators/tests/event_based_risk_test.py]: Done

File: openquake/risklib/scientific.py (hunk 5 above)
Comment (micheles) [inline openquake/risklib/scientific.py]: Done

File: openquake/risklib/scientific.py (hunk 6 above)
Comment (micheles) [inline openquake/risklib/scientific.py]: Done

**Pull request merged**